deps/redis/crc64.c

```diff
@@ -122,6 +122,20 @@ uint64_t crc64(uint64_t crc, const unsigned char *s, uint64_t l) {
     return crcspeed64native(crc64_table, crc, (void *) s, l);
 }
 
+/* Not sure if it is really required to use mutex here, but just in case */
```

Comment:
I saw late but yes, we need a mutex here. Strictly speaking, there is still a race condition. For the outer if check, a thread might observe `crcInitalized = 1` even before `crc64_init()` is completed.

To solve the issue, we can get rid of the outer if check or make `crcInitalized` an atomic value. This is a known problem, please see https://en.wikipedia.org/wiki/Double-checked_locking

Reply:
I am not sure if you followed the entire thread around it. 
There was an issue with my original implementation that set the flag before calling crc64_init(), but I still argue that:
> ... it is generally safe for two different threads to write the same value to a variable (int this case it is a table) as long as there are no race conditions or other synchronization issues involved. Writing the same value(s) to a variable simultaneously won't cause data corruption or unexpected behavior by itself.

In any way, to be on the safe I modified it to use mutex with double-check-locking for optimization.

Reply:
I found an interesting article about [double-check-locking](https://www.aristeia.com/Papers/DDJ_Jul_Aug_2004_revised.pdf). Its conclusion is to use memory barriers. It is like falling down the rabbit hole.

I will go back to square one and claim that simply calling `crc64_init()` is lock-free programming withtout any risks.

> For example, if crc64_init() clears the table first or updates a table value twice, 

You can look at the code of `crc64_init`,  crc init doesn't memset the table and only makes write of constant values. And this not goanna change anytime soon.

The buttom line is maybe I should revert `crc64_init_thread_safe()` and call `crc64_init()` instead.

Reply:
If there is one place I have the courage to write lock-free code, is during intialization alone and with function
that not goanna change. CRC tables are initialized this way for ages. For any other function I would have agree with you.

Reply:
FYI: https://github.com/redis/redis/pull/11988
i.e. i'm not certain things are always that simple and i do think it can change.

i rather not make assumptions, and either make a global init API with restrictions, or use safe locking without any tricks.

Reply:
Yes, but he didn't touched `crc64_init()` :) Cool.

Ok, I got your point. I will create a PR with the following change:

```
void crc64_init_thread_safe(void) {
    static int crcInitalized = 0;
    static pthread_mutex_t mutex = PTHREAD_MUTEX_INITIALIZER;

    pthread_mutex_lock(&mutex);
    if (!crcInitalized) {
        crc64_init();
        crcInitalized = 1;
    }
    pthread_mutex_unlock(&mutex);
}
```
It supposed to be fine even in case of reordering or multiprocessor machine. 

Thanks